### ISSUE TITLE
Look into performance improvements

### ISSUE THREAD

**Issue by alanxz** [alanxz/rabbitmq-c#262]
The rather vague request here is to improve the performance of the library. One facet might be reducing the number of `send()`/`writev()` syscalls as brought up here: https://groups.google.com/forum/#!topic/rabbitmq-users/cwwnlxx781g

Action items for this issue:

- [ ] Define the various scenarios that performance is important and should be measured
- [ ] Come up with some test-cases that can be used to test this performance release-to-release
- [ ] Create improvements in rabbitmq-c itself.

**Comment by alanxz** [comment]
I've had a chance to do some performance testing using the amqp_producer.c example program:

Running on a 64-bit Linux 3.13.0-49-generic #83-Ubuntu SMP kernel, 2xIntel(R) Core(TM)2 Duo CPU     T9400  @ 2.53GHz (This is running inside a VM).
RabbitMQ broker: RabbitMQ 3.5.1 on Erlang R16B03, running on the localhost.

librabbitmq-c v0.6.0
**With a 256-byte message size**:
PRODUCER - Message count: 1000000
Total time, milliseconds: 20099
Overall messages-per-second: 49752.1

real    0m21.062s
user    0m0.468s
sys     0m1.967s

**With a 96KB message size**:
PRODUCER - Message count: 100000
Total time, milliseconds: 17255
Overall messages-per-second: 5795.16

real    0m17.307s
user    0m0.113s
sys     0m7.006s

With code currently in master that eliminates 2 of the send calls per message (so 3 send calls per message):
**With a 256-byte message size**:
PRODUCER - Message count: 1000000
Total time, milliseconds: 20137
Overall messages-per-second: 49657.7

real    0m21.065s
user    0m0.406s
sys     0m1.263s

**With a 96KB message size**:
PRODUCER - Message count: 100000
Total time, milliseconds: 20647
Overall messages-per-second: 4843.17

real    0m20.677s
user    0m2.176s
sys     0m6.650s

For short messages the performance of the two implementations seems more or less equal (or at least its hard to pick out from the noise).  For longer messages the use of the 5-send() call seems to have slightly better performance (~13% faster, with ~0.5s more spent in system calls).  In both cases the limiting factor was not rabbitmq-c, but the broker, which pegged showed quite a bit of CPU usage (as I would expect).

**Comment by isj4** [comment]
An alternative is to use TCP_CORK under Linux. My main interest is the bandwidth used by AMQP on expensive network links. In my case amqp_basic_publish() results in 3 TCP packets sent and 3 TCP ACKs received. The following (hackish) patch reduces it to 1 TCP packet sent and 1 TCP ACK received thus reducing the IP+TCP overhead (roughly 208 bytes saved per basic publish). Presumably this also reduces the load on the reciever slightly.

diff -r 729d0f2acc1c librabbitmq/amqp_api.c
--- a/librabbitmq/amqp_api.c    Tue May 19 16:30:27 2015 +0200
+++ b/librabbitmq/amqp_api.c    Wed May 20 16:29:33 2015 +0200
@@ -119,9 +119,17 @@
   m.immediate = immediate;
   m.ticket = 0;
 
+  int old_cork_state = 0;
+  socklen_t old_cork_state_size = sizeof(old_cork_state);
+  getsockopt(state->sockfd, IPPROTO_TCP, TCP_CORK, &old_cork_state, &old_cork_state_size);
+  int new_cork_state = 1;
+  setsockopt(state->sockfd, IPPROTO_TCP, TCP_CORK, &new_cork_state, sizeof(new_cork_state));
+  
   res = amqp_send_method(state, channel, AMQP_BASIC_PUBLISH_METHOD, &m);
-  if (res < 0)
+  if (res < 0) {
+    setsockopt(state->sockfd, IPPROTO_TCP, TCP_CORK, &old_cork_state, sizeof(old_cork_state));
     return res;
+  }
 
   if (properties == NULL) {
     memset(&default_properties, 0, sizeof(default_properties));
@@ -135,8 +143,10 @@
   f.payload.properties.decoded = (void *) properties;
 
   res = amqp_send_frame(state, &f);
-  if (res < 0)
+  if (res < 0) {
+    setsockopt(state->sockfd, IPPROTO_TCP, TCP_CORK, &old_cork_state, sizeof(old_cork_state));
     return res;
+  }
 
   body_offset = 0;
   while (body_offset < body.len) {
@@ -156,10 +166,14 @@
 
     body_offset += f.payload.body_fragment.len;
     res = amqp_send_frame(state, &f);
-    if (res < 0)
+    if (res < 0) {
+      setsockopt(state->sockfd, IPPROTO_TCP, TCP_CORK, &old_cork_state, sizeof(old_cork_state));
       return res;
+    }
   }
 
+setsockopt(state->sockfd, IPPROTO_TCP, TCP_CORK, &old_cork_state, sizeof(old_cork_state));
+
   return 0;
 }

**Comment by alanxz** [comment]
That's an interesting solution. I'll give it a try on my end see what kind of performance difference it makes.

**Comment by alanxz** [comment]
After looking at it a bit more, `TCP_CORK` (or equivalent) is a viable way to improve performance by reducing the number of packets transmitted per publish.

Some implementation notes:
- This needs to be a part of the transport layer so that transports like SSL can be handled correctly.
- `MSG_MORE` is probably more efficient on Linux (cuts out a pair of `setsockopt()` calls)
- `TCP_NOPUSH` works on FreeBSD (probably Mac OS X too). This is a `setsockopt()` call.
- `MSG_PARTIAL` may do the same thing on Win32 systems.
- This could possibly be used to gain back the (small amount) of performance lost when we switched away from `writev()` in favor of copying to an output buffer.

**Comment by isj4** [comment]
Yes, my hackish patch ignores those complexities. I had forgotten about MSG_MORE. In general, TCP_CORK is used by upper logic when it doesn't control the send() calls, and MSG_MORE is used when the logic can control the parameter in the send() calls.

It seems to me that amqp_api.c() needs to tell the lower layer when a series of coherent writes is starting and when they are finished. The lower layer can then use MSG_MORE/TCP_CORK/TCP_NOPUSH and buffering as appropriate/available.

If I remember correctly, MSG_PARTIAL under Windows is only supported by message-oriented protocols (UDP/SCTP/DCCP/...) so output buffering is the only available mechanism on that platform.

**Comment by mohammednaseef** [comment]
Yes, MSG_MORE will supersede the performance improvements obtained using TCP_CORK /TCP_NODELAY. I have already tested this during the rabbitmq-users discussion mentioned by Alan on top.

Alan, since the new version is using writev() , will not the MSG_MORE be eliminated anyways ?

**Comment by alanxz** [comment]
While there is a small amount of overhead in extra syscalls, the
performance overhead of the syscall itself is small in comparison to
network write (we're talking ~100ns on most modern CPUs, vs ~1ms to send
packets over the wire)

The new version will likely not use writev, as there's not a good way to
pass MSG_NOSIGNAL to writev, and no good way to disable signals on a per
socket level.

Additionally there's significant code complexity to using writev with non
blocking sockets that doesn't justify the relatively minor performance
overhead of extra syscalls

**Comment by isj4** [comment]
sendmsg() supports using both vector-writes and flags. But I agree that the setsockopt() syscall overhead is neglible, so the goal should be to optimize network performance while keeping rabbitmq-c library easy to maintain.

**Comment by alanxz** [comment]
From my comments in the PR #272:

Evaluating this running on Ubuntu 14.04.02 running in a VM on Mac OS X 10.11, talking to RabbitMQ 3.5.2 running different machine Mac OS X 10.11 connected over a 100mb ethernet connection.

Comparing PR #272 against the v0.6.0 tag using the amqp_producer example, using 256 byte and 96 KB message sizes.

Looking at the packets sent using `tcpdump` for the first few messages, the packets are sent as I would expect for both methods, however, after the first few message are sent, the kernel start combining messages into larger packets so the performance of the two methods ends up being nearly identical (both in terms of msg/s and packets sent).

So granted this is just one performance datapoint, but its unclear to me that this will have a large positive impact on sustained message publish performance (for periodic bursty message publishing this might be a clear win).

@isj4: how are you measuring publish performance? Would you care to share some performance numbers comparing v0.6.0 and whats in #272?

**Comment by isj4** [comment]
I measured performance with tcpdump/wireshark.
My interest is not millions of messages per second. My interest is the network bandwidth used for small messages, maybe 10-100 messages per minute. The actual payload of the basic publish is 170-353 bytes.

Per-packet on-the-wire overhead:
Ethernet header: 14 bytes
IP header: 20 bytes
TCP header: 32 bytes
= 66 bytes.

In the default rabbitmq-c which unconditionally disables Nagle and issues multiple write/send/writev this results in a basic publish causing 3 packets sent and 3 acks received:
Basic.Publish: 141 bytes (we are using routing headers)
Content-Header: 106 bytes
Content-Body: 244 bytes (170 bytes actual payload)
3 acks of each 32 bytes
= 587 bytes total

With my hackish patch with TCP_CORK the message is sent as a coherent whole and a single tcp ack is received:
basic-publish+content-header+content-body: 359 bytes
1 ack: 66 bytes
= 425 bytes

= 27% reduction in traffic.

Note: our traffic is sent through an openvpn tunnel which strips off the ethernet header, but then adds additional overhead per packet (packet tag, hmac-sha1, IV, seq, ...) so the overhead in the production system is probably more.

So making sure that a basic-publish is sent as a coherent whole is a good thing for small messages.

I can provide pcapng files for both original rabbitmq-c and with my hack.

**Comment by alanxz** [comment]
Understood, thanks for the explanation. I believe that once I can get #272 cleaned up and pushed to master this will improve things for your use case (feel free to try it out).

**Comment by mohammednaseef** [comment]
I suppose there is room for both the approaches to co-exist, one favouring the throughput and another the bandwidth. Perhaps there should be a different amqp publish method/ additional parameter which indicates the mode the user wants to send the messages.

@alanxz: Is there a possibility of having a common amqp header and footer for a burst of messages. Many a times the user may want to send say, 1000 messages which are part of one operation. All these messages will have the same routing key and are sent to the same exchange. If the user can send the header only once and send a end of operation message after all the 1000 messages, we can gain on bandwidth and possibly, latency also.

**Comment by alanxz** [comment]
No. Not without changing the way AMQP works. (This doesn't stop you from putting more data in one message though).

**Comment by alanxz** [comment]
#272 is now in the master branch. This will improve things for those using non-SSL sockets on Linux and BSDs that support TCP_NOPUSH.

I have #273 open to do a more significant refactoring of how basic.publish works using something like `writev()` to better support OSs that don't have things like MSG_MORE and TCP_NOPUSH and OpenSSL where its not easy to create a maintainable buffering mechanism using the current API.

**Comment by isj4** [comment]
Just wanted to say that I have switched to version 0.8.0 and it works beautifully - each basic.publish results in 1 TCP packet and one ACK. Thanks!
